### ISSUE TITLE
Dynamic Dims not getting produced right.

### ISSUE THREAD

**Issue by cptspacemanspiff** [pytorch/pytorch#146046]
### 🐛 Describe the bug

So, I have been using torch.export, but ran into the annoying and confusing issue of the dynamic dims do not in all cases produce dynamic dims.

Specifically, I encountered this when dealing with sliced assignments in python, the relative size of the dynamic dimentions compared to the example makes a difference:

Example for dynamic dim [1 : 10] 
* If example is size 1, it fails and errors out with an inferred constant.
* If example is size 9, it works
* If example is size 10, it fails, **_does not error out_**, but the graph does not have any dynamic dimentions (presumably because internally it inferred constant was 10).

A reproduction python code below shows this:

```python
from torch.export import export, Dim
import torch


class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()

        self.register_buffer("cache", torch.zeros(10))

    def forward(self, x):
        self.cache[0 : x.shape[0]] = x
        return x

max_size = 10
min_size = 1
    
dim = Dim(name="dim", min=min_size, max=max_size)

module = Foo()

# This works and is correct:
# exported_not_max_size
# ExportedProgram:
#     class GraphModule(torch.nn.Module):
#         def forward(self, b_cache: "f32[10]", x: "f32[s0]"):
#              # 
#             sym_size_int_2: "Sym(s0)" = torch.ops.aten.sym_size.int(x, 0)
            
#              # File: /home/nlong/execu-tools/python/tests/export_dynamic_size_bug.py:12 in forward, code: self.cache[0 : x.shape[0]] = x
#             slice_1: "f32[s0]" = torch.ops.aten.slice.Tensor(b_cache, 0, 0, sym_size_int_2);  b_cache = sym_size_int_2 = None
#             copy_: "f32[s0]" = torch.ops.aten.copy_.default(slice_1, x);  slice_1 = copy_ = None
#             return (x,)
            
# Graph signature: ExportGraphSignature(input_specs=[InputSpec(kind=<InputKind.BUFFER: 3>, arg=TensorArgument(name='b_cache'), target='cache', persistent=True), InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='x'), target=None, persistent=None)], output_specs=[OutputSpec(kind=<OutputKind.USER_OUTPUT: 1>, arg=TensorArgument(name='x'), target=None)])
# Range constraints: {s0: VR[1, 10]}
exported_not_max_size = export(
    module,
    (
        torch.zeros(
            max_size - 1,
        ),
    ),
    dynamic_shapes={"x": {0: dim}},
)

# This works, but is incorrect:
# exported_max_size
# ExportedProgram:
#     class GraphModule(torch.nn.Module):
#         def forward(self, b_cache: "f32[10]", x: "f32[10]"):
#              # File: /home/nlong/execu-tools/python/tests/export_dynamic_size_bug.py:12 in forward, code: self.cache[0 : x.shape[0]] = x
#             copy_: "f32[10]" = torch.ops.aten.copy_.default(b_cache, x);  b_cache = copy_ = None
#             return (x,)
            
# Graph signature: ExportGraphSignature(input_specs=[InputSpec(kind=<InputKind.BUFFER: 3>, arg=TensorArgument(name='b_cache'), target='cache', persistent=True), InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='x'), target=None, persistent=None)], output_specs=[OutputSpec(kind=<OutputKind.USER_OUTPUT: 1>, arg=TensorArgument(name='x'), target=None)])
# Range constraints: {}
exported_max_size = export(
    module,
    (
        torch.zeros(
            max_size,
        ),
    ),
    dynamic_shapes={"x": {0: dim}},
)

# This fails with the message: 
# Not all values of dim = L['x'].size()[0] in the specified range dim <= 10 are valid because dim was inferred to be a constant (1).
# exported_min_size = export(
#     module,
#     (
#         torch.zeros(
#             min_size,
#         ),
#     ),
#     dynamic_shapes={"x": {0: dim}},
# )



print("---------------------------------------")
print("exported_not_max_size")
print(exported_not_max_size)
print("---------------------------------------")
print("exported_max_size")
print(exported_max_size)
print("---------------------------------------")
# print("exported_min_size")
# print(exported_min_size)


```

The silent failure is the biggest issue, that being said a better error message with the inferred constant linking to a blogpost on the topic would have saved me a lot of time.

Not sure if it is valid, but my ideal, would have a bright red warning saying this may fail if your dimensions are 0/1/max. Though I realize max may be a special case for this sliced copy.

### Versions

torch==2.6.0.dev20250104+cpu

**Comment by avikchaudhuri** [comment]
Thanks for reporting, looking...

**Comment by avikchaudhuri** [comment]
Interesting tidbit here: if you use `strict=False` you get a constraint violation error in both cases: the first because of a guard `s0 != 10` and the second because of a guard `s0 == 10`. 

According to @pianpwk the graph gets specialized or not based on a contiguity check during slicing. For some reason Dynamo eats the guard emitted by this condition but non-strict doesn't.

With `Dim.AUTO` and `strict=False` you'd replicate the behavior here.

cc @bobrenjc93 @laithsakka @ezyang

**Comment by avikchaudhuri** [comment]
Separately, it is always the case that we specialize on size `1` because of https://docs.google.com/document/d/16VPOa3d-Liikf48teAOmxLc92rgvJdfosIy-yoT38Io/edit?fbclid=IwAR3HNwmmexcitV0pbZm_x1a4ykdXZ9th_eJWK-3hBtVgKnrkmemz6Pm5jRQ&tab=t.0#heading=h.ez923tomjvyk

**Comment by avikchaudhuri** [comment]
@pianpwk What would be the fix here? 

Also, if I make `max_size = 9` then the VR of `s0` is `[1,9]` but if I use `Dim.AUTO` then I get `[2, 10]` and never specialize the graph. The overall design does seem a bit confusing here, we should try to rationalize it in our heads.

**Comment by pianpwk** [comment]
My understanding is that this is what the size-oblivious upper bound is for https://github.com/pytorch/pytorch/pull/144867, it just doesn't apply to backed SymInts right now?
https://github.com/pytorch/pytorch/blob/eb029fba13d0829471e35d77adbe83d38e6a6912/torch/fx/experimental/symbolic_shapes.py#L1975-L1986

**Comment by jansel** [comment]
My agent was unable to reproduce this on b034e24e3d5191102a4a55d91d4548bc5688e85b. Reopen if you see this on the latest PyTorch nightly
